File: scripts/migrate-mem0-to-byterover.ts
#!/usr/bin/env tsx
/**
 * Migrate memories from Mem0 Platform and OpenMemory to ByteRover CLI
 * 
 * Usage:
 *   # From Mem0 Platform
 *   tsx scripts/migrate-mem0-to-byterover.ts --source mem0 --api-key YOUR_MEM0_API_KEY
 * 
 *   # From OpenMemory (hosted)
 *   tsx scripts/migrate-mem0-to-byterover.ts --source openmemory --api-key YOUR_OPENMEMORY_API_KEY
 * 
 *   # From OpenMemory (local)
 *   tsx scripts/migrate-mem0-to-byterover.ts --source openmemory-local --url http://localhost:8765
 * 
 *   # Dry run (preview without importing)
 *   tsx scripts/migrate-mem0-to-byterover.ts --source mem0 --api-key KEY --dry-run
 */

import { spawnSync } from 'child_process';
import { writeFileSync, mkdirSync } from 'fs';
import { join, dirname } from 'path';

interface Mem0Memory {
    id: string;
    memory: string;
    created_at: string;
    updated_at: string;
    user_id?: string;
    agent_id?: string;
    categories?: string[];
    metadata?: Record<string, any>;
}

interface OpenMemoryItem {
    id: string;
    content: string;
    timestamp: string;
    metadata?: Record<string, any>;
}

interface MigrationConfig {
    source: 'mem0' | 'openmemory' | 'openmemory-local';
    apiKey?: string;
    url?: string;
    dryRun: boolean;
    filters?: Record<string, any>;
    batchSize: number;
}

class Mem0Migrator {
    private config: MigrationConfig;
    private baseUrl = 'https://api.mem0.ai/v2';

    constructor(config: MigrationConfig) {
        this.config = config;
    }

    async fetchMem0Memories(page = 1): Promise<{ memories: Mem0Memory[]; hasMore: boolean }> {
        const response = await fetch(`${this.baseUrl}/memories/`, {
            method: 'POST',
            headers: {
                'Authorization': `Token ${this.config.apiKey}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                filters: this.config.filters || {},
                page,
                page_size: this.config.batchSize,
            }),
        });

        if (!response.ok) {
            throw new Error(`Mem0 API error: ${response.status} ${response.statusText}`);
        }

        const data = await response.json();
        return {
            memories: data.results || [],
            hasMore: data.results?.length === this.config.batchSize,
        };
    }

    async fetchOpenMemoryItems(): Promise<OpenMemoryItem[]> {
        // For hosted OpenMemory, use their API
        const url = this.config.url || 'https://api.openmemory.dev';

        const response = await fetch(`${url}/memories`, {
            headers: {
                'Authorization': `Bearer ${this.config.apiKey}`,
            },
        });

        if (!response.ok) {
            throw new Error(`OpenMemory API error: ${response.status} ${response.statusText}`);
        }

        return await response.json();
    }

    async exportAllMemories(): Promise<Array<{ content: string; section: string; metadata: any }>> {
        console.log(`📥 Exporting memories from ${this.config.source}...`);

        const allMemories: Array<{ content: string; section: string; metadata: any }> = [];

        if (this.config.source === 'mem0') {
            let page = 1;
            let hasMore = true;

            while (hasMore) {
                console.log(`  Fetching page ${page}...`);
                const { memories, hasMore: more } = await this.fetchMem0Memories(page);

                for (const mem of memories) {
                    allMemories.push({
                        content: mem.memory,
                        section: this.categorizeMemory(mem),
                        metadata: {
                            originalId: mem.id,
                            source: 'mem0',
                            createdAt: mem.created_at,
                            updatedAt: mem.updated_at,
                            userId: mem.user_id,
                            agentId: mem.agent_id,
                            categories: mem.categories || [],
                            tags: ['migrated-from-mem0', ...(mem.categories || [])],
                        },
                    });
                }

                hasMore = more;
                page++;
            }
        } else {
            // OpenMemory
            const items = await this.fetchOpenMemoryItems();

            for (const item of items) {
                allMemories.push({
                    content: item.content,
                    section: 'Lessons Learned',
                    metadata: {
                        originalId: item.id,
                        source: 'openmemory',
                        timestamp: item.timestamp,
                        tags: ['migrated-from-openmemory'],
                    },
                });
            }
        }

        console.log(`✅ Exported ${allMemories.length} memories`);
        return allMemories;
    }

    categorizeMemory(mem: Mem0Memory): string {
        // Map Mem0 categories to ByteRover sections
        const categories = mem.categories || [];

        if (categories.some(c => c.includes('error') || c.includes('bug'))) {
            return 'Common Errors';
        }
        if (categories.some(c => c.includes('best') || c.includes('practice'))) {
            return 'Best Practices';
        }
        if (categories.some(c => c.includes('architecture') || c.includes('design'))) {
            return 'Architecture';
        }
        if (categories.some(c => c.includes('test'))) {
            return 'Testing';
        }
        if (categories.some(c => c.includes('strategy') || c.includes('approach'))) {
            return 'Strategies';
        }

        return 'Lessons Learned';
    }

    async importToByteRover(memories: Array<{ content: string; section: string; metadata: any }>) {
        console.log(`\n📤 Importing ${memories.length} memories to ByteRover...`);

        if (this.config.dryRun) {
            console.log('\n🔍 DRY RUN - Preview of memories to import:\n');
            memories.slice(0, 5).forEach((mem, idx) => {
                console.log(`${idx + 1}. [${mem.section}] ${mem.content.substring(0, 100)}...`);
            });
            console.log(`\n... and ${memories.length - 5} more`);
            return;
        }

        let successCount = 0;
        let errorCount = 0;

        for (const mem of memories) {
            try {
                // Use brv add command with safe argument passing (prevents command injection)
                const args = ['add', '--section', mem.section, '--content', mem.content];
                const result = spawnSync('brv', args, { stdio: 'pipe' });

                if (result.status !== 0) {
                    throw new Error(`brv add failed: ${result.stderr?.toString()}`);
                }

                successCount++;

                if (successCount % 10 === 0) {
                    console.log(`  Imported ${successCount}/${memories.length}...`);
                }
            } catch (error) {
                console.error(`  ❌ Failed to import: ${mem.content.substring(0, 50)}...`);
                errorCount++;
            }
        }

        console.log(`\n✅ Migration complete!`);
        console.log(`   Success: ${successCount}`);
        console.log(`   Errors: ${errorCount}`);

        // Save backup with timestamp to avoid overwriting previous backups
        const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
        const backupFilename = `migration-backup-${timestamp}.json`;
        const backupPath = join(process.cwd(), '.brv', backupFilename);
        mkdirSync(dirname(backupPath), { recursive: true });
        writeFileSync(backupPath, JSON.stringify(memories, null, 2));
        console.log(`\n💾 Backup saved to: ${backupPath}`);
    }

    async migrate() {
        try {
            const memories = await this.exportAllMemories();
            await this.importToByteRover(memories);
        } catch (error) {
            console.error('❌ Migration failed:', error);
            process.exit(1);
        }
    }
}

// CLI
const args = process.argv.slice(2);
const config: MigrationConfig = {
    source: 'mem0',
    dryRun: false,
    batchSize: 100,
};

for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
        case '--source':
            config.source = args[++i] as any;
            break;
        case '--api-key':
            config.apiKey = args[++i];
            break;
        case '--url':
            config.url = args[++i];
            break;
        case '--dry-run':
            config.dryRun = true;
            break;
        case '--batch-size':
            config.batchSize = parseInt(args[++i]);
            break;
        case '--help':
            console.log(`
Migrate memories from Mem0/OpenMemory to ByteRover CLI

Usage:
  tsx scripts/migrate-mem0-to-byterover.ts [options]

Options:
  --source <type>       Source type: mem0, openmemory, openmemory-local
  --api-key <key>       API key for Mem0 or OpenMemory
  --url <url>           Custom URL for local OpenMemory (default: http://localhost:8765)
  --dry-run             Preview without importing
  --batch-size <n>      Batch size for pagination (default: 100)
  --help                Show this help

Examples:
  # From Mem0 Platform
  tsx scripts/migrate-mem0-to-byterover.ts --source mem0 --api-key YOUR_KEY

  # From OpenMemory (hosted)
  tsx scripts/migrate-mem0-to-byterover.ts --source openmemory --api-key YOUR_KEY

  # Dry run
  tsx scripts/migrate-mem0-to-byterover.ts --source mem0 --api-key YOUR_KEY --dry-run
      `);
            process.exit(0);
    }
}

if (!config.apiKey && config.source !== 'openmemory-local') {
    console.error('❌ --api-key is required');
    process.exit(1);
}

const migrator = new Mem0Migrator(config);
migrator.migrate();

File: scripts/migrate_mem0_to_byterover.py
#!/usr/bin/env python3
"""
Migrate memories from Mem0 Platform and OpenMemory to ByteRover CLI

Usage:
    # From Mem0 Platform
    python scripts/migrate_mem0_to_byterover.py --source mem0 --api-key YOUR_MEM0_API_KEY

    # From OpenMemory
    python scripts/migrate_mem0_to_byterover.py --source openmemory --api-key YOUR_KEY

    # Dry run
    python scripts/migrate_mem0_to_byterover.py --source mem0 --api-key KEY --dry-run
"""
# ruff: noqa: T201

import argparse
import json
import subprocess
import sys
from pathlib import Path
from typing import Any

import requests


class Mem0Migrator:
    def __init__(
        self, source: str, api_key: str | None = None, url: str | None = None, dry_run: bool = False
    ):
        self.source = source
        self.api_key = api_key
        self.url = url or "http://localhost:8765"
        self.dry_run = dry_run
        self.mem0_base_url = "https://api.mem0.ai/v2"

    def fetch_mem0_memories(self, page: int = 1, page_size: int = 100) -> dict[str, Any]:
        """Fetch memories from Mem0 Platform API"""
        headers = {
            "Authorization": f"Token {self.api_key}",
            "Content-Type": "application/json",
        }

        # Use v2 endpoint with pagination in query params and filters in body
        # Based on official SDK: client.post("/v2/memories/", json=params, params=query_params)
        # Filters are required - use wildcard to match all
        payload = {"filters": {"user_id": "*"}}  # Wildcard to match all users
        query_params = {"page": page, "page_size": page_size}

        response = requests.post(
            f"{self.mem0_base_url}/memories/",
            headers=headers,
            json=payload,
            params=query_params,
        )
        response.raise_for_status()
        return response.json()

    def fetch_openmemory_items(self) -> list[dict[str, Any]]:
        """Fetch memories from OpenMemory"""
        headers = {}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        response = requests.get(f"{self.url}/memories", headers=headers)
        response.raise_for_status()
        return response.json()

    def categorize_memory(self, memory: dict[str, Any]) -> str:
        """Map Mem0 categories to ByteRover sections"""
        categories = memory.get("categories")
        if not categories or not isinstance(categories, list):
            return "Lessons Learned"

        # Map categories to sections
        category_map = {
            ("error", "bug"): "Common Errors",
            ("best", "practice"): "Best Practices",
            ("architecture", "design"): "Architecture",
            ("test",): "Testing",
            ("strategy", "approach"): "Strategies",
        }

        for cat in categories:
            cat_lower = cat.lower()
            for keywords, section in category_map.items():
                if any(keyword in cat_lower for keyword in keywords):
                    return section

        return "Lessons Learned"

    def export_all_memories(self) -> list[dict[str, Any]]:
        """Export all memories from source"""
        print(f"📥 Exporting memories from {self.source}...")
        all_memories = []

        if self.source == "mem0":
            page = 1
            while True:
                print(f"  Fetching page {page}...")
                data = self.fetch_mem0_memories(page=page)
                memories = data.get("results", [])

                if not memories:
                    break

                for mem in memories:
                    categories = mem.get("categories")
                    categories_list = (
                        categories if categories and isinstance(categories, list) else []
                    )

                    all_memories.append(
                        {
                            "content": mem["memory"],
                            "section": self.categorize_memory(mem),
                            "metadata": {
                                "originalId": mem["id"],
                                "source": "mem0",
                                "createdAt": mem.get("created_at"),
                                "updatedAt": mem.get("updated_at"),
                                "userId": mem.get("user_id"),
                                "agentId": mem.get("agent_id"),
                                "categories": categories_list,
                                "tags": ["migrated-from-mem0", *categories_list],
                            },
                        }
                    )

                # Check if there are more pages
                if len(memories) < 100:
                    break
                page += 1

        elif self.source in ["openmemory", "openmemory-local"]:
            items = self.fetch_openmemory_items()

            for item in items:
                all_memories.append(
                    {
                        "content": item.get("content", item.get("memory", "")),
                        "section": "Lessons Learned",
                        "metadata": {
                            "originalId": item.get("id"),
                            "source": "openmemory",
                            "timestamp": item.get("timestamp", item.get("created_at")),
                            "tags": ["migrated-from-openmemory"],
                        },
                    }
                )

        print(f"✅ Exported {len(all_memories)} memories")
        return all_memories

    def import_to_byterover(self, memories: list[dict[str, Any]]):
        """Import memories to ByteRover using brv CLI"""
        print(f"\n📤 Importing {len(memories)} memories to ByteRover...")

        if self.dry_run:
            print("\n🔍 DRY RUN - Preview of memories to import:\n")
            for idx, mem in enumerate(memories[:5], 1):
                content_preview = mem["content"][:100]
                print(f"{idx}. [{mem['section']}] {content_preview}...")
            print(f"\n... and {len(memories) - 5} more")
            return

        success_count = 0
        error_count = 0

        for mem in memories:
            try:
                # Escape quotes in content
                content = mem["content"].replace('"', '\\"')
                section = mem["section"]

                # Run brv add command
                cmd = ["brv", "add", "--section", section, "--content", content]
                subprocess.run(cmd, check=True, capture_output=True, text=True)

                success_count += 1
                if success_count % 10 == 0:
                    print(f"  Imported {success_count}/{len(memories)}...")

            except OSError as e:
                print(
                    f"  ❌ Error: Failed to execute 'brv' CLI. Is it installed and in your PATH? Details: {e}"
                )
                error_count += 1
            except subprocess.CalledProcessError as e:
                print(f"  ❌ Failed to import: {mem['content'][:50]}...")
                print(f"     Error: {e.stderr}")
                error_count += 1

        print("\n✅ Migration complete!")
        print(f"   Success: {success_count}")
        print(f"   Errors: {error_count}")

        # Save backup with timestamp to avoid overwriting previous backups
        from datetime import datetime

        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        backup_path = Path.cwd() / ".brv" / f"migration-backup_{timestamp}.json"
        backup_path.parent.mkdir(parents=True, exist_ok=True)
        with open(backup_path, "w") as f:
            json.dump(memories, f, indent=2)
        print(f"\n💾 Backup saved to: {backup_path}")

    def migrate(self):
        """Run the full migration"""
        try:
            memories = self.export_all_memories()
            self.import_to_byterover(memories)
        except requests.exceptions.HTTPError as e:
            print(f"❌ Migration failed: HTTP {e.response.status_code}", file=sys.stderr)
            print(f"Response: {e.response.text}", file=sys.stderr)
            print("\nTroubleshooting:")
            print("  1. Verify your API key is correct")
            print("  2. Check if you have access to the Mem0 API")
            print(
                f"  3. Try the check script first: python scripts/check-mem0-memories.py --source {self.source} --api-key YOUR_KEY"
            )
            sys.exit(1)
        except Exception as e:
            import traceback  # noqa: PLC0415

            print(f"❌ Migration failed: {e}", file=sys.stderr)
            traceback.print_exc()
            sys.exit(1)


def main():
    parser = argparse.ArgumentParser(
        description="Migrate memories from Mem0/OpenMemory to ByteRover CLI"
    )
    parser.add_argument(
        "--source",
        required=True,
        choices=["mem0", "openmemory", "openmemory-local"],
        help="Source type",
    )
    parser.add_argument("--api-key", help="API key for Mem0 or OpenMemory")
    parser.add_argument(
        "--url",
        default="http://localhost:8765",
        help="Custom URL for local OpenMemory",
    )
    parser.add_argument("--dry-run", action="store_true", help="Preview without importing")

    args = parser.parse_args()

    if not args.api_key and args.source != "openmemory-local":
        print("❌ --api-key is required", file=sys.stderr)
        sys.exit(1)

    migrator = Mem0Migrator(
        source=args.source,
        api_key=args.api_key,
        url=args.url,
        dry_run=args.dry_run,
    )
    migrator.migrate()


if __name__ == "__main__":
    main()
